service/matching/version_sets.go

```diff
@@ -0,0 +1,196 @@
+	// First find if the targeted version is already in the sets
+	targetedVersion := extractTargetedVersion(req)
+	targetSetIx, versionInSetIx := findVersion(existingData, targetedVersion)
+
+	if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_AddNewVersionIdInNewDefaultSet); ok {
+		// If it's not already in the sets, add it as the new default set
+		if targetSetIx != -1 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		existingData.VersionSets = append(existingData.GetVersionSets(), &taskqueuepb.CompatibleVersionSet{
+			Versions: []string{targetedVersion},
+		})
+	} else if op, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_AddNewCompatibleVersion_); ok {
+		compatVer := op.AddNewCompatibleVersion.GetExistingCompatibleVersion()
+		compatSetIx, _ := findVersion(existingData, compatVer)
+		if compatSetIx == -1 {
+			return serviceerror.NewNotFound(
+				fmt.Sprintf("targeted compatible_version %v not found", compatVer))
+		}
+		if targetSetIx != -1 {
+			// If the version does exist, this operation can't do anything meaningful, but we can fail if the user
+			// says the version is now compatible with some different set.
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		// If the version doesn't exist, add it to the compatible set
+		existingData.VersionSets[compatSetIx].Versions =
+			append(existingData.VersionSets[compatSetIx].Versions, targetedVersion)
+		if op.AddNewCompatibleVersion.GetMakeSetDefault() {
+			makeDefaultSet(existingData, compatSetIx)
+		}
+	} else if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_PromoteSetByVersionId); ok {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeDefaultSet(existingData, targetSetIx)
+	} else if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_PromoteVersionIdWithinSet); ok {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeVersionInSetDefault(existingData, targetSetIx, versionInSetIx)
+	}
+
+	return nil
+}
+
+func extractTargetedVersion(req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) string {
+	if req.GetAddNewCompatibleVersion() != nil {
+		return req.GetAddNewCompatibleVersion().GetNewVersionId()
+	} else if req.GetPromoteSetByVersionId() != "" {
+		return req.GetPromoteSetByVersionId()
+	} else if req.GetPromoteVersionIdWithinSet() != "" {
+		return req.GetPromoteVersionIdWithinSet()
+	}
+	return req.GetAddNewVersionIdInNewDefaultSet()
+}
+
+// Finds the version in the version sets, returning (set index, index within that set)
+// Returns -1, -1 if not found.
+func findVersion(data *persistence.VersioningData, id string) (int, int) {
```

Comment:
nit, but consider naming it `buildID` everywhere so it's clear what type of ID is being used here?

(server codebase style is to use all caps for `ID` in names, contrary to Go style. yes, even though proto-generated code uses `Id`. sorry)